cli/upload.py

```diff
@@ -177,55 +194,72 @@ def _wait_for_upload(procs: list) -> bool:
                 if p.returncode != 0:
                     errored = True
 
+                    for line in p.stderr:
+                        print(f"\t:\t{line}")
+
+                    break
+
             errline = p.stderr.readline()
 
             # skipping "large file" warnings
             if errline.strip() in _IGNORED_WARN_LINES:
                 continue
 
-            if errline and len(errline) > 2:  # skipping '* ' spinner lines
+            if errline.strip() and len(errline) > 2:  # skipping '* ' spinner lines
 
                 # changing output from always "[0/1 files]" - due to manual
                 # parallelization of gsutil to "[x/y files]"
                 if errline.split("]", 1)[0].endswith("/1 files"):
                     errline = errline.split("]", 1)[1]
                 print(
-                    f"[{len(finished)}/{len(procs)} done] {i+1}: {p.args[-2]}: "
-                    + errline,
-                    end="",
+                    f"[{len(finished)}/{total} done] {i+1}: " + errline, end="",
                 )
 
-    print(f"[{len(finished)}/{len(procs)} done]")
     return errored
 
 
+# default from `gsutil -m` but maybe better to load from env
+MAX_GSUTIL_PARALLEL_PROCESS = 12
+
+
 def _gsutil_assay_upload(upload_info: api.UploadInfo, xlsx: str):
     """
     Upload local assay data to GCS using gsutil.
     """
 
-    procs = _start_procs(upload_info, xlsx)
+    upload_pairs = _compose_file_mapping(upload_info, xlsx)
+
+    it = _start_procs(upload_pairs)
+    procs = []
+    while True:
 
-    err = _wait_for_upload(procs)
+        try:
+            for _ in range(min(len(procs) or 1, MAX_GSUTIL_PARALLEL_PROCESS)):
+                p = next(it)
+                procs.append(p)
```

Comment:
This code seems to handle the upload process batching logic. Can you provide comments explaining it?